source/link/linker.cpp

```diff
@@ -16,1183 +16,277 @@
-
-      prev_j = j;
-
-      spv::StorageClass storage_type = static_cast<spv::StorageClass>(GetWord(binary, j + 3u));
-      if (!IsTypeOrConstantOpcode(opcode) ||
-          (opcode == spv::Op::OpVariable &&
-           storage_type == spv::StorageClass::Function))
-        break;
-
-      j += word_count;
-
-      // Remove imported variables
-      spv::Id tmp = static_cast<spv::Id>(GetWord(binary, prev_j + 2u));
-      if (opcode == spv::Op::OpVariable &&
-         link_table.find(tmp) != link_table.end()) {
-        std::memcpy(merged_binary + merged_length, binary + index,
-                    (prev_j - index) * sizeof(spirv_word));
-        merged_length += prev_j - index;
-        index = j;
-      }
+  // Remove declarations of imported global variables
+  for (auto i = linkedModule->types_values_begin(); i != linkedModule->types_values_end();) {
+    if (i->NumOperands() < 2u) {
+      ++i;
+      continue;
     }
-    if (j == length)
-      prev_j = j;
-
-    std::memcpy(merged_binary + merged_length, binary + index,
-                (prev_j - index) * sizeof(spirv_word));
-    merged_length += prev_j - index;
-    indices[i] = j;
+    const auto variable_id = i->GetSingleWordOperand(1u);
+    if (imports.find(variable_id) == imports.end()) {
+      ++i;
+      continue;
+    }
+    i = i.Erase();
   }
 
-  // Skip function declarations
-  for (uint32_t i = 0u; i < num_binaries; ++i) {
-    const uint32_t* const binary = binaries[i].data();
-    const size_t length = binaries[i].size();
-    const uint32_t index = indices[i];
-    uint32_t j = index;
-    uint32_t function_begin = j;
-    while (j < length) {
-      uint32_t word_count;
-      const spv::Op opcode = ParseOpcode(binary, j, word_count);
+  // TODO Remove imported types
 
-      if (opcode == spv::Op::OpFunction)
-        function_begin = j;
+  // Remove debug instructions for imported declarations
+  for (auto i = linkedModule->debug_begin(); i != linkedModule->debug_end();) {
+    bool should_remove = false;
+    i->ForEachInId([&imports,&should_remove](const uint32_t* id){
+        should_remove |= imports.find(*id) != imports.end();
 
-      // This was not a prototype: revert to the beginning of the function
-      if (opcode != spv::Op::OpFunction && opcode != spv::Op::OpFunctionParameter &&
-          opcode != spv::Op::OpFunctionEnd) {
-         j = function_begin;
-        break;
-      }
-
-      j += word_count;
+    });
+    if (should_remove) {
+      i = i.Erase();
+      continue;
     }
-
-    indices[i] = j;
+    ++i;
   }
 
-  // Copy function defintions
-  for (uint32_t i = 0u; i < num_binaries; ++i) {
-    const uint32_t* const binary = binaries[i].data();
-    const size_t length = binaries[i].size();
-    const uint32_t index = indices[i];
-    std::memcpy(merged_binary + merged_length, binary + index,
-                (length - index) * sizeof(spirv_word));
-    merged_length += length - index;
-  }
+  // TODO Remove duplicate decorations
```

Comment:
Explain more?  These would be decorations that are the same but on both an import ID and export ID that are really the same object?